archive/100-theorems-list/57_herons_formula.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2020 Matt Kempster. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matt Kempster
+-/
+import geometry.euclidean.triangle
+
+/-!
+# Freek № 57: Heron's Formula
+
+This file proves Theorem 57 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/),
+also known as Heron's formula, which gives the area of a triangle based only on its three sides'
+lengths.
+
+## References
+
+* https://en.wikipedia.org/wiki/Herons_formula
+
+-/
+
+open euclidean_geometry
+open_locale real euclidean_geometry
+
+notation `√` := real.sqrt
+
+variables {V : Type*} {P : Type*} [inner_product_space ℝ V] [metric_space P]
+    [normed_add_torsor V P]
+include V
+
+/-- The sine of a real number `x` is the square root of one minus the cosine squared of `x`. -/
+lemma sin_eq_sqrt_one_minus_cos_sq (x : ℝ) (hl : x ≥ 0) (hu : x ≤ π) :
+  real.sin x = √ (1 - (real.cos x)^2) :=
+begin
+  rw ← real.sqrt_sqr (real.sin_nonneg_of_nonneg_of_le_pi hl hu),
+  exact congr_arg real.sqrt (real.sin_square x),
+end
+
+lemma rearrange_cos_rule (a b c d : ℝ) (a_nonzero : a ≠ 0) (b_nonzero : b ≠ 0):
+  c * c = a * a + b * b - 2 * a * b * d → d = (a * a + b * b - c * c) / (2 * a * b) :=
+begin
+  intro hyp,
+  simp only [*, sub_sub_cancel],
+  field_simp,
+  ring,
+end
+
+/-- Heron's formula: The area of a triangle with side lengths `a`, `b`, and `c` is
+  `√( s * (s - a) * (s - b) * (s - c) )` where `s = ( a + b + c) / 2` is the semiperimeter.
+  We show this by equating this formula to `a * b * sin γ`, where `γ` is the angle opposite
+  the side `c`.
+ -/
+theorem heron {p1 p2 p3 : P} (h2 : p1 ≠ p2) (h3 : p1 ≠ p3) (h4 : p3 ≠ p2) :
+  1/2 * dist p1 p2 * dist p3 p2 * real.sin (∠ p1 p2 p3) =
+    √ ( ( ( dist p1 p2 + dist p3 p2 + dist p1 p3 ) / 2 ) *
+        ( ( ( dist p1 p2 + dist p3 p2 + dist p1 p3 ) / 2 ) - dist p1 p2 ) *
+        ( ( ( dist p1 p2 + dist p3 p2 + dist p1 p3 ) / 2 ) - dist p3 p2 ) *
+        ( ( ( dist p1 p2 + dist p3 p2 + dist p1 p3 ) / 2 ) - dist p1 p3 ) ) :=
```

Comment:
I believe the style guide dictates against orphaning parentheses as such.

Reply:
What does orphaning mean? I thought the style guide was referring to parens on their own lines, so I must have misinterpreted.

Reply:
Hmm, maybe it does? Tbh I had always thought orphaning parentheses meant leaving them unbalanced, but I had interpreted the style guide to be referring to adding spaces around them. Regardless, as Bryan says, we generally don't put spaces around parentheses.

Reply:
The style guide doesn't mention not putting spaces around parentheses explicitly (maybe it should), but the code samples there don't show any spaces around parentheses and I think most code in mathlib avoids extra spaces like this as well.

Reply:
Ah, perhaps orphaning means putting spaces around them. Very well, removing them now.